### PR TITLE
#23 Implement AD LDAP authentication

### LINKED ISSUE #23: Simple username+pwd auth against AD
This issue is an individual issue broken down from #8.

Logically depends on #21.

Here we assume a simple AD integration:
 - config changes to allow multiple modes = aside from built-in user(s) to offer AD-mode, too
 - AD URL(s) will be given in config
 - auth endpoint with user/pwd may be reused

### PR DESCRIPTION
**Pull request by jakipatryk**
- added me to codeowners/developers
- added AD LDAP auth provider
- made `email` in `User` model an `Option` as service accounts don't have emails

DI of AD LDAP config to be done as part of separate issue (as this should be provider-agnostic work).

Closes #23

### CODE REVIEW

File: service/src/main/scala/za/co/absa/logingw/rest/SecurityConfig.scala [hunk 1]
```diff
@@ -53,6 +56,27 @@ class SecurityConfig {
...
+
+    // TODO: take which providers and in which order to use from config
+    authenticationManagerBuilder
+      // if it is not null, on auth failure infinite recursion happens
+      .parentAuthenticationManager(null)
+      .authenticationProvider(
+        new ActiveDirectoryLDAPAuthenticationProvider(adLDAPConfig)
+      )
+      .authenticationProvider(
+        new DummyAuthenticationProvider()
```
Comment (dk1844) [inline service/src/main/scala/za/co/absa/logingw/rest/SecurityConfig.scala:77]: wow, can you really chain the providers this way?

File: service/src/main/scala/za/co/absa/logingw/model/User.scala [hunk 2]
```diff
@@ -16,4 +16,4 @@
 
 package za.co.absa.logingw.model
 
-case class User(name: String, email: String, groups: Seq[String])
+case class User(name: String, email: Option[String], groups: Seq[String])
```
Comment (dk1844) [inline service/src/main/scala/za/co/absa/logingw/model/User.scala:19]: why this change btw?

**Review by dk1844 — COMMENTED** [review]
👍

File: service/src/main/scala/za/co/absa/logingw/rest/SecurityConfig.scala (hunk 1 above)
Comment (jakipatryk) [reply service/src/main/scala/za/co/absa/logingw/rest/SecurityConfig.scala:77]: yup, thats cool

File: service/src/main/scala/za/co/absa/logingw/model/User.scala (hunk 2 above)
Comment (jakipatryk) [reply service/src/main/scala/za/co/absa/logingw/model/User.scala:19]: service accounts don't have email

File: README.md [hunk 3]
```diff
@@ -30,6 +30,19 @@ Swagger doc site is available at `http://localhost:8080/swagger-ui.html`
...
+### ActiveDirectoryLDAPAuthenticationProvider
+Uses LDAP(s) to authenticate user in Active Directory and to fetch groups that this user belongs to.
+
+Requires `ActiveDirectoryLDAPConfig(domain: String, url: String, searchFilter: String)`.
+#### Tips
+##### How to obtain certificate for LDAPS
+1. Run `openssl s_client -connect <ldaps_host>:<ldaps_port>`.
+2. Copy the part starting with `-----BEGIN CERTIFICATE-----` and ending with `-----END CERTIFICATE-----`. 
+3. Create file `ldapcert.pem` and paste content from (2) there.
+4. Run `keytool -import -file ldapcert.pem -alias ldaps -keystore <path_to_jdk>/Contents/Home/jre/lib/security/cacerts -storepass <password>`.
```
Comment (dk1844) [inline README.md:43]:
```suggestion
4. Run `keytool -import -file ldapcert.pem -alias ldaps -keystore <path_to_jdk>/jre/lib/security/cacerts -storepass <password>`.
```

For me, the path looks like this ⬆️ 
For completeness' sake, I would probably briefly note that the default password for `cacerts` is `changeit`.

**Review by dk1844 — COMMENTED** [review]
Otherwise LGTM.

File: service/src/main/scala/za/co/absa/logingw/rest/SecurityConfig.scala [hunk 4]
```diff
@@ -53,6 +56,27 @@ class SecurityConfig {
...
+    val authenticationManagerBuilder = http.getSharedObject(classOf[AuthenticationManagerBuilder])
+
+    // TODO make it autowired but only if in confid AD LDAP provider is set up
+    val adLDAPConfig = ActiveDirectoryLDAPConfig(
+      "some.domain.com",
+      "ldaps://some.domain.com:636/",
+      "(samaccountname={1})"
+    )
+
+    // TODO: take which providers and in which order to use from config
```
[the remark this answers is not in the document]
Comment (jakipatryk) [reply service/src/main/scala/za/co/absa/logingw/rest/SecurityConfig.scala]: Done

File: README.md (hunk 3 above)
Comment (jakipatryk) [reply README.md]: Makes sense, changed.

**Review by dk1844 — APPROVED** [review]
LGTM now (just read the code)

**Pull request merged**